text/0000-windows-subsystem.md

```diff
@@ -0,0 +1,150 @@
+parse those linker arguments.
+
+A possible future extension would be to add additional command-line options to
+`rustc` (and in turn, `Cargo.toml`) to specify the subsystem directly. `rustc`
+would automatically translate this into the correct linker arguments for
+whichever linker is actually being used.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Additional platform-specific code.
+- The difficulty of manually calling the rust initialization code is potentially
+  a more general problem, and this only solves a specific (if common) case.
+- This is a breaking change for any crates which already export a `WinMain`
+  symbol. It is likely that only executable crates would export this symbol,
+  so the knock-on effect on crate dependencies should be non-existent.
+
+  A possible work-around for this is described below.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Emit either `WinMain` or `main` from `libstd` based on `cfg` options.
+
+  This has the advantage of not requiring changes to `rustc`, but is something
+  of a non-starter since it requires a version of `libstd` for each subsystem.
+
+- Emit either `WinMain` or `main` from `rustc` based on `cfg` options.
+
+  This would not require different versions of `libstd`, but it would require
+  recompiling all other crates depending on the value of the `cfg` option.
+
+- Emit either `WinMain` or `main` from `rustc` based on a new command line
+  option.
+
+  Assuming the command line option need only be specified when compiling the
+  executable itself, the dependencies would not need to be recompiled were the
+  subsystem to change.
+
+  Choosing to emit one or the other means that the compiler and linker must
+  agree on the subsystem, or else you'll get linker errors. If `rustc` only
+  specified a `subsystem` to the linker if the option is passed, this would be
+  a fully backwards compatible change.
+
+  A compiler option is probably desirable in addition to this RFC, but it will
+  require bike-shedding on the new command line interface, and changes to rustc
+  to be able to pass on the correct linker flags.
+
+  A similar option would need to be added to `Cargo.toml` to make usage as simple
+  as possible.
+
+- Add a `subsystem` function to determine which subsystem was used at runtime.
+
+  The `WinMain` function would first set an internal flag, and only then
+  delegate to the `main` function.
+
+  A function would be added to `std::os::windows`:
+
+  `fn subsystem() -> &'static str`
+
+  This would check the value of the internal flag, and return either `WINDOWS` or
+  `CONSOLE` depending on which entry point was actually used.
+
+  The `subsystem` function could be used to eg. redirect logging to a file if
+  the program is being run on the `WINDOWS` subsystem. However, it would return
+  an incorrect value if the initialization was skipped, such as if used as a
+  library from an executable written in another language.
```

Comment:
I'm not sure I understand this alternative in terms of how it might replace the detailed design section of this RFC, wouldn't this be a nice-to-have regardless of how this RFC turns out?